aptos-move/mvhashmap/src/versioned_data.rs

```diff
@@ -115,14 +133,121 @@ impl<V: TransactionWrite> Default for VersionedValue<V> {
+        if let Some((_, entry)) = self
+            .versioned_map
+            .range(..=ShiftedTxnIndex::new(txn_idx))
+            .next_back()
+        {
+            // Non-exchanged format is default validation failure.
+            if let EntryCell::ResourceWrite {
+                incarnation: _,
+                value_with_layout,
+                dependencies,
+            } = &entry.value
+            {
+                // Take dependencies above txn_idx
+                affected_deps = dependencies.lock().split_off(&(txn_idx + 1, 0));
+                if !affected_deps.is_empty() {
+                    if let ValueWithLayout::Exchanged(entry_value, entry_maybe_layout) =
+                        value_with_layout
+                    {
+                        still_valid = compare_values_and_layouts::<ONLY_COMPARE_METADATA, V>(
+                            entry_value,
+                            data,
+                            entry_maybe_layout.as_ref(),
+                            maybe_layout.as_ref(),
+                        );
+                    }
+                }
+            }
+        }
+        (affected_deps, still_valid)
+    }
+
+    /// Handle dependencies from a removed entry by validating against the next (lower) entry.
+    /// The caller MUST ensure that the entry at txn_idx has been removed from versioned_map
+    /// before calling this method. If the lower entry does not exist, all dependencies are
+    /// considered invalidated. A re-execution of the read can then create a baseline sentinel
+    /// entry, if needed. Data is the value that was stored in the removed entry.
+    fn handle_removed_dependencies<const ONLY_COMPARE_METADATA: bool>(
+        &mut self,
+        txn_idx: TxnIndex,
+        mut dependencies: BTreeSet<(TxnIndex, Incarnation)>,
+        data: &Arc<V>,
+        maybe_layout: &Option<Arc<MoveTypeLayout>>,
+    ) -> BTreeSet<(TxnIndex, Incarnation)> {
+        // If we have dependencies and a next (lower) entry exists, validate against it.
+        if !dependencies.is_empty() {
+            if let Some((idx, next_entry)) = self
+                .versioned_map
+                .range(..=ShiftedTxnIndex::new(txn_idx))
+                .next_back()
+            {
+                assert_ne!(
+                    idx.idx(),
+                    Ok(txn_idx),
+                    "Entry at txn_idx must be removed before calling handle_removed_dependencies"
+                );
+
+                // Non-exchanged format is default validation failure.
+                if let EntryCell::ResourceWrite {
+                    incarnation: _,
+                    value_with_layout: ValueWithLayout::Exchanged(entry_value, entry_maybe_layout),
+                    dependencies: next_deps,
+                } = &next_entry.value
+                {
+                    let still_valid = compare_values_and_layouts::<ONLY_COMPARE_METADATA, V>(
```

Comment:
Why not if compare_values_and_layouts ?

Reply:
it's like a comment, shows that the outcome of that call determines whether something is still valid